pkg/apk/impl/repo.go

```diff
@@ -0,0 +1,443 @@
+	// - /bin/sh   - this means, "I need package busybox"
+	// - name      - this means, "I need package 'name'"
+	// - so:<lib>  - this means, "find the package that provides so:<lib>"
+	// - cmd:<cmd> - this means, "find the package that provides cmd:<cmd>"
+	for _, dep := range ref.Dependencies {
+		var newDep string
+		switch {
+		case dep == "/bin/sh":
+			newDep = "busybox"
+		case strings.HasPrefix(dep, "so:"), strings.HasPrefix(dep, "cmd:"):
+			provider, ok := pkgProvidesMap[dep]
+			if !ok {
+				return nil, nil, fmt.Errorf("could not find package that provides %s for %s", dep, pkg)
+			}
+			newDep = provider
+		case strings.HasPrefix(dep, "!"):
+			conflicts = append(conflicts, dep[1:])
+			continue
+		default:
+			newDep = dep
+		}
+		// this package might be pinned to a version
+		name, version, compare := resolvePackageNameVersion(newDep)
+		depPkg, ok := pkgNameMap[name]
+		if !ok {
+			// it might be an actual package name, or it might be an alias, i.e. another package that provides this one
+			provider, ok := pkgProvidesMap[name]
+			if !ok {
+				return nil, nil, fmt.Errorf("could not find package %s", newDep)
+			}
+			depPkg, ok = pkgNameMap[provider]
+			if !ok {
+				return nil, nil, fmt.Errorf("required package %s is provided by package %s, but could not find package %s", newDep, provider, provider)
+			}
+		}
+		if err := compareVersion(depPkg.Version, version, compare); err != nil {
+			return nil, nil, fmt.Errorf("mismatched version for package %s in indexes: %w", newDep, err)
+		}
+		dependencies = append(dependencies, depPkg)
+		// and then recurse to its children
+		subDeps, confs, err := getPackageDependencies(depPkg.Name, pkgNameMap, pkgProvidesMap)
+		if err != nil {
+			return nil, nil, err
+		}
+		dependencies = append(dependencies, subDeps...)
+		conflicts = append(conflicts, confs...)
+	}
+	return dependencies, conflicts, nil
+}
+
+type versionDependency int
+
+const (
+	versionNone versionDependency = iota
+	versionEqual
+	versionGreater
+	versionLess
+	versionGreaterEqual
+	versionLessEqual
+)
+
+func resolvePackageNameVersion(name string) (string, string, versionDependency) {
+	parts := strings.SplitN(name, "=", 2)
+	if len(parts) == 2 {
+		// we have an equal
+		if strings.HasSuffix(parts[0], ">") {
+			return strings.TrimSuffix(parts[0], ">"), parts[1], versionGreaterEqual
+		}
+		if strings.HasSuffix(parts[0], "<") {
+			return strings.TrimSuffix(parts[0], "<"), parts[1], versionLessEqual
+		}
+		return parts[0], parts[1], versionEqual
+	}
+	parts = strings.SplitN(name, ">", 2)
+	if len(parts) == 2 {
+		// we have a greater than but not equal
+		return strings.TrimSuffix(parts[0], ">"), parts[1], versionGreater
+	}
+	parts = strings.SplitN(name, "<", 2)
+	if len(parts) == 2 {
+		// we have a less than but not equal
+		return strings.TrimSuffix(parts[0], "<"), parts[1], versionLess
+	}
+	// we have no =, < or >, so we just return the name
+	return name, "", versionNone
+}
+
+// compare semver versions
```

Comment:
apk versions aren't semver, but rather derived from Gentoo versions.  See https://dev.gentoo.org/~ulm/pms/head/pms.html#x1-250003.2 for details about the version grammar and comparison rules.

Reply:
Ah you know, I was wondering about that. I didn't bother digging too deeply, figuring that since it is all in a single function, I can fix it to whatever is "correct" (after I find the definition of "correct") later. This helps greatly, thank you. I will update it.

Reply:
Completely revamped to follow the logic on that link (and includes the link), and lots of test cases for it in a table.